File: agent-kernel/internal/protocol/contracts.go
package protocol

import (
	"encoding/hex"
	"errors"
	"fmt"
	"strings"
	"time"
)

const (
	ProtocolVersion = "resume-agent/v1"
	ProposalVersion = "agent-action-proposal/v1"
	ToolsetVersion  = "resume-readonly-tools/v1"
	ResultVersion   = "resume-screening/v1"
	PolicyVersion   = "django-policy-gate/v1"
)

type Pin struct {
	PinID               string `json:"pin_id"`
	KernelBuild         string `json:"kernel_build"`
	ProtocolVersion     string `json:"protocol_version"`
	ToolsetVersion      string `json:"toolset_version"`
	ResultSchemaVersion string `json:"result_schema_version"`
	PolicyVersion       string `json:"policy_version"`
	PromptVersion       string `json:"prompt_version"`
	ModelConfigRevision string `json:"model_config_revision"`
}

type Constraints struct {
	WorkflowRevision int64    `json:"workflow_revision"`
	VolunteerRank    int      `json:"volunteer_rank"`
	Policies         []string `json:"policies"`
}

type CandidateReference struct {
	HighestMajor string `json:"highest_major"`
}

type Volunteer struct {
	PositionName string `json:"position_name"`
}

type JobContext struct {
	Entity           string   `json:"entity"`
	PublicName       string   `json:"public_name"`
	PositionName     string   `json:"position_name"`
	Category         string   `json:"category"`
	JobFamily        string   `json:"job_family"`
	Location         string   `json:"location"`
	RequiredMajors   []string `json:"required_majors"`
	Responsibilities string   `json:"responsibilities"`
	DepartmentName   string   `json:"department_name"`
}

type ResumeContent struct {
	Checksum string `json:"checksum"`
	Text     string `json:"text"`
}

type ModelConfig struct {
	APIStyle           string  `json:"api_style"`
	BaseURL            string  `json:"base_url"`
	ModelName          string  `json:"model_name"`
	StructuredOutput   string  `json:"structured_output_mode"`
	TimeoutSeconds     float64 `json:"timeout_seconds"`
	RetryCount         int     `json:"retry_count"`
	InsecureSkipVerify bool    `json:"insecure_skip_verify"`
}

type Budget struct {
	MaxTurns           int `json:"max_turns"`
	MaxToolCalls       int `json:"max_tool_calls"`
	MaxDurationSeconds int `json:"max_duration_seconds"`
}

type CaseEnvelopeV1 struct {
	ProtocolVersion  string             `json:"protocol_version"`
	TaskID           string             `json:"task_id"`
	IdempotencyKey   string             `json:"idempotency_key"`
	Pin              Pin                `json:"pin"`
	Constraints      Constraints        `json:"constraints"`
	Candidate        CandidateReference `json:"candidate_reference"`
	CurrentVolunteer Volunteer          `json:"current_volunteer"`
	CurrentJob       JobContext         `json:"current_job"`
	Resume           ResumeContent      `json:"resume"`
	Instructions     string             `json:"instructions"`
	Model            ModelConfig        `json:"model"`
	Budget           Budget             `json:"budget"`
}

func (e CaseEnvelopeV1) Validate() error {
	if e.ProtocolVersion != ProtocolVersion {
		return fmt.Errorf("unsupported protocol_version %q", e.ProtocolVersion)
	}
	if strings.TrimSpace(e.TaskID) == "" || strings.TrimSpace(e.IdempotencyKey) == "" {
		return errors.New("task_id and idempotency_key are required")
	}
	if e.Pin.ProtocolVersion != ProtocolVersion || e.Pin.ToolsetVersion != ToolsetVersion || e.Pin.ResultSchemaVersion != ResultVersion || e.Pin.PolicyVersion != PolicyVersion {
		return errors.New("pinned protocol, toolset, result schema, or policy is incompatible")
	}
	if strings.TrimSpace(e.Pin.PinID) == "" || strings.TrimSpace(e.Pin.KernelBuild) == "" || strings.TrimSpace(e.Pin.PromptVersion) == "" || strings.TrimSpace(e.Pin.ModelConfigRevision) == "" {
		return errors.New("pin_id, kernel_build, prompt_version, and model_config_revision are required")
	}
	if e.Constraints.WorkflowRevision < 0 || e.Constraints.VolunteerRank < 1 || e.Constraints.VolunteerRank > 4 || len(e.Constraints.Policies) == 0 {
		return errors.New("workflow_revision, volunteer_rank, and policies are invalid")
	}
	if len(e.Resume.Checksum) != 64 {
		return errors.New("resume checksum must be a 64-character SHA-256")
	}
	if _, err := hex.DecodeString(e.Resume.Checksum); err != nil {
		return errors.New("resume checksum must be hexadecimal")
	}
	if strings.TrimSpace(e.Resume.Text) == "" {
		return errors.New("resume checksum and text are required")
	}
	if len(e.Resume.Text) > 60_000 {
		return errors.New("resume text exceeds 60000 characters")
	}
	if strings.TrimSpace(e.CurrentJob.Responsibilities) == "" {
		return errors.New("current job responsibilities are required")
	}
	if strings.TrimSpace(e.CurrentVolunteer.PositionName) == "" || strings.TrimSpace(e.CurrentJob.PositionName) == "" || strings.TrimSpace(e.Instructions) == "" {
		return errors.New("current volunteer, current job, and instructions are required")
	}
	if e.Model.APIStyle != "chat_json" && e.Model.APIStyle != "responses" {
		return errors.New("unsupported model api_style")
	}
	if strings.TrimSpace(e.Model.BaseURL) == "" || strings.TrimSpace(e.Model.ModelName) == "" {
		return errors.New("model base_url and model_name are required")
	}
	if e.Model.TimeoutSeconds <= 0 || e.Model.TimeoutSeconds > 1800 {
		return errors.New("timeout_seconds must be in (0, 1800]")
	}
	if e.Model.RetryCount < 0 || e.Model.RetryCount > 5 {
		return errors.New("retry_count must be between 0 and 5")
	}
	if e.Budget.MaxTurns < 1 || e.Budget.MaxTurns > 8 {
		return errors.New("max_turns must be between 1 and 8")
	}
	if e.Budget.MaxToolCalls < 1 || e.Budget.MaxToolCalls > 16 {
		return errors.New("max_tool_calls must be between 1 and 16")
	}
	if e.Budget.MaxDurationSeconds < 1 || e.Budget.MaxDurationSeconds > 1800 {
		return errors.New("max_duration_seconds must be between 1 and 1800")
	}
	return nil
}

type ExperienceItem struct {
	Name        string `json:"name"`
	Role        string `json:"role"`
	Period      string `json:"period"`
	Description string `json:"description"`
	Evidence    string `json:"evidence"`
}

type EducationItem struct {
	SchoolName string `json:"school_name"`
	Degree     string `json:"degree"`
	Major      string `json:"major"`
	Period     string `json:"period"`
	Evidence   string `json:"evidence"`
}

type ScoreBreakdown struct {
	MajorMatch         float64 `json:"major_match"`
	SkillsMatch        float64 `json:"skills_match"`
	ExperienceEvidence float64 `json:"experience_evidence"`
	JobRequirement     float64 `json:"job_requirement"`
	ResumeQuality      float64 `json:"resume_quality"`
}

func (s ScoreBreakdown) Validate() error {
	values := []float64{s.MajorMatch, s.SkillsMatch, s.ExperienceEvidence, s.JobRequirement, s.ResumeQuality}
	for _, value := range values {
		if value < 0 || value > 1 {
			return errors.New("score breakdown values must be between 0 and 1")
		}
	}
	return nil
}

type ResumeProfile struct {
	MajorDirection string           `json:"major_direction"`
	Educations     []EducationItem  `json:"educations"`
	Projects       []ExperienceItem `json:"projects"`
	Internships    []ExperienceItem `json:"internships"`
	Skills         []string         `json:"skills"`
	Certificates   []string         `json:"certificates"`
	Summary        string           `json:"summary"`
	RiskFlags      []string         `json:"risk_flags"`
}

type Decision struct {
	Recommendation         string         `json:"recommendation"`
	ScoreBreakdown         ScoreBreakdown `json:"score_breakdown"`
	Summary                string         `json:"summary"`
	Reason                 string         `json:"reason"`
	Evidence               []string       `json:"evidence"`
	Risks                  []string       `json:"risks"`
	AISpecialistMatch      bool           `json:"ai_specialist_match"`
	AISpecialistConfidence float64        `json:"ai_specialist_confidence"`
	AISpecialistEvidence   []string       `json:"ai_specialist_evidence"`
}

type ScreeningOutput struct {
	Profile  ResumeProfile `json:"profile"`
	Decision Decision      `json:"decision"`
}

func (o ScreeningOutput) Validate() error {
	switch o.Decision.Recommendation {
	case "dispatch", "review", "archive":
	default:
		return errors.New("invalid recommendation")
	}
	if err := o.Decision.ScoreBreakdown.Validate(); err != nil {
		return err
	}
	if o.Decision.AISpecialistConfidence < 0 || o.Decision.AISpecialistConfidence > 1 {
		return errors.New("ai_specialist_confidence must be between 0 and 1")
	}
	return nil
}

type ToolCall struct {
	ID        string         `json:"id"`
	Name      string         `json:"name"`
	Arguments map[string]any `json:"arguments"`
}

type TurnFinal struct {
	Action     string          `json:"action"`
	Evaluation ScreeningOutput `json:"evaluation"`
}

type AgentTurnOutput struct {
	Kind      string     `json:"kind"`
	ToolCalls []ToolCall `json:"tool_calls"`
	Final     *TurnFinal `json:"final"`
}

type ToolTrace struct {
	Name       string `json:"name"`
	Status     string `json:"status"`
	DurationMS int64  `json:"duration_ms"`
	ItemCount  int    `json:"item_count"`
}

type SafeTrace struct {
	TraceID       string      `json:"trace_id"`
	KernelBuild   string      `json:"kernel_build"`
	StartedAt     time.Time   `json:"started_at"`
	FinishedAt    time.Time   `json:"finished_at"`
	Turns         int         `json:"turns"`
	ToolCallCount int         `json:"tool_call_count"`
	ToolCalls     []ToolTrace `json:"tool_calls"`
	InputTokens   int         `json:"input_tokens"`
	OutputTokens  int         `json:"output_tokens"`
	Status        string      `json:"status"`
}

type EvaluationError struct {
	Cause error
	Trace SafeTrace
}

func (e *EvaluationError) Error() string {
	return e.Cause.Error()
}

func (e *EvaluationError) Unwrap() error {
	return e.Cause
}

type AgentActionProposalV1 struct {
	ProposalVersion string          `json:"proposal_version"`
	TaskID          string          `json:"task_id"`
	PinID           string          `json:"pin_id"`
	Action          string          `json:"action"`
	Evaluation      ScreeningOutput `json:"evaluation"`
	Trace           SafeTrace       `json:"safe_trace"`
}

File: backend/apps/pipeline/agent_kernel/contracts.py
"""Agent Kernel v1 协议；这里只描述不可变输入和无写权限建议。"""

from __future__ import annotations

import hashlib
import json
from datetime import datetime
from typing import Literal

from django.conf import settings
from django.core.exceptions import ObjectDoesNotExist
from pydantic import BaseModel, ConfigDict, Field

from apps.pipeline import ai_config
from apps.pipeline.ai import prompt_harness
from apps.pipeline.ai.schemas import ResumeScreeningOutput


PROTOCOL_VERSION = "resume-agent/v1"
PROPOSAL_VERSION = "agent-action-proposal/v1"
TOOLSET_VERSION = "resume-readonly-tools/v1"
RESULT_SCHEMA_VERSION = "resume-screening/v1"
POLICY_VERSION = "django-policy-gate/v1"


class StrictModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class KernelPinV1(StrictModel):
    pin_id: str
    kernel_build: str
    protocol_version: Literal[PROTOCOL_VERSION] = PROTOCOL_VERSION
    toolset_version: Literal[TOOLSET_VERSION] = TOOLSET_VERSION
    result_schema_version: Literal[RESULT_SCHEMA_VERSION] = RESULT_SCHEMA_VERSION
    policy_version: Literal[POLICY_VERSION] = POLICY_VERSION
    prompt_version: str
    model_config_revision: str


class CaseConstraintsV1(StrictModel):
    workflow_revision: int = Field(ge=0)
    volunteer_rank: int = Field(ge=1, le=4)
    policies: list[str]


class CandidateReferenceV1(StrictModel):
    highest_major: str = ""


class CurrentVolunteerV1(StrictModel):
    position_name: str


class FixedJobContextV1(StrictModel):
    entity: str = ""
    public_name: str = ""
    position_name: str = ""
    category: str = ""
    job_family: str = ""
    location: str = ""
    required_majors: list[str] = Field(default_factory=list)
    responsibilities: str = Field(min_length=1, max_length=12_000)
    department_name: str = ""


class ResumeContentV1(StrictModel):
    checksum: str = Field(min_length=64, max_length=64)
    text: str = Field(min_length=1, max_length=60_000)


class KernelModelConfigV1(StrictModel):
    api_style: Literal["chat_json", "responses"]
    base_url: str
    model_name: str
    structured_output_mode: str
    timeout_seconds: float = Field(gt=0, le=1800)
    retry_count: int = Field(ge=0, le=5)
    insecure_skip_verify: bool = False


class KernelBudgetV1(StrictModel):
    max_turns: int = Field(default=6, ge=1, le=8)
    max_tool_calls: int = Field(default=12, ge=1, le=16)
    max_duration_seconds: int = Field(default=600, ge=1, le=1800)


class CaseEnvelopeV1(StrictModel):
    protocol_version: Literal[PROTOCOL_VERSION] = PROTOCOL_VERSION
    task_id: str
    idempotency_key: str
    pin: KernelPinV1
    constraints: CaseConstraintsV1
    candidate_reference: CandidateReferenceV1
    current_volunteer: CurrentVolunteerV1
    current_job: FixedJobContextV1
    resume: ResumeContentV1
    instructions: str = Field(min_length=1, max_length=32_000)
    model: KernelModelConfigV1
    budget: KernelBudgetV1 = Field(default_factory=KernelBudgetV1)


class ToolTraceV1(StrictModel):
    name: str
    status: Literal["success", "rejected"]
    duration_ms: int = Field(ge=0)
    item_count: int = Field(ge=0)


class SafeTraceV1(StrictModel):
    trace_id: str
    kernel_build: str
    started_at: datetime
    finished_at: datetime
    turns: int = Field(ge=1, le=8)
    tool_call_count: int = Field(ge=0, le=16)
    tool_calls: list[ToolTraceV1]
    input_tokens: int = Field(ge=0)
    output_tokens: int = Field(ge=0)
    status: Literal["completed"]


class AgentActionProposalV1(StrictModel):
    proposal_version: Literal[PROPOSAL_VERSION]
    task_id: str
    pin_id: str
    action: Literal["dispatch", "review", "archive"]
    evaluation: ResumeScreeningOutput
    safe_trace: SafeTraceV1


def _pin_id(payload):
    encoded = json.dumps(
        payload,
        ensure_ascii=False,
        sort_keys=True,
        separators=(",", ":"),
    ).encode("utf-8")
    return hashlib.sha256(encoded).hexdigest()


def build_runtime_pin(model_config):
    pin_payload = {
        "kernel_build": getattr(settings, "AGENT_KERNEL_BUILD", "dev"),
        "protocol_version": PROTOCOL_VERSION,
        "toolset_version": TOOLSET_VERSION,
        "result_schema_version": RESULT_SCHEMA_VERSION,
        "policy_version": POLICY_VERSION,
        "prompt_version": model_config.prompt_version,
        "model_config_revision": ai_config.current_ai_connection_fingerprint(),
    }
    return KernelPinV1(pin_id=_pin_id(pin_payload), **pin_payload)


def build_envelope(prepared, *, processing_run_id=None, pin=None):
    """从 ORM 对象构造无数据库 ID 的不可变 Kernel 输入。"""

    try:
        workflow = prepared.resume.candidate.workflow
    except ObjectDoesNotExist:
        workflow = None
    workflow_revision = max(0, int(getattr(workflow, "revision", 0) or 0))
    volunteer_rank = int(prepared.resume.volunteer_rank or 1)
    pin = pin or build_runtime_pin(prepared.model_config)
    task_id = (
        f"run-{processing_run_id}-{prepared.checksum[:12]}"
        if processing_run_id is not None
        else f"adhoc-{prepared.checksum[:16]}"
    )
    idempotency_key = _pin_id(
        {
            "task_id": task_id,
            "resume_checksum": prepared.checksum,
            "workflow_revision": workflow_revision,
            "pin_id": pin.pin_id,
        }
    )
    current_job = dict(prepared.job_context)
    current_job["department_name"] = (
        prepared.department.name if prepared.department else ""
    )
    return CaseEnvelopeV1(
        task_id=task_id,
        idempotency_key=idempotency_key,
        pin=pin,
        constraints=CaseConstraintsV1(
            workflow_revision=workflow_revision,
            volunteer_rank=volunteer_rank,
            policies=[
                "只处理当前有效志愿",
                "岗位与二级部门引用已经固定，不得替换",
                "学历与院校准入结果已经由 Django Policy Gate 确认",
                "最终业务写入必须由 Django Policy/Executor 完成",
            ],
        ),
        candidate_reference=CandidateReferenceV1(
            highest_major=prepared.resume.candidate.highest_major or ""
        ),
        current_volunteer=CurrentVolunteerV1(
            position_name=prepared.resume.position_name or ""
        ),
        current_job=FixedJobContextV1.model_validate(current_job),
        resume=ResumeContentV1(
            checksum=prepared.checksum,
            text=prepared.text[:60_000],
        ),
        instructions=prompt_harness.build_screening_prompt(
            prompt_harness.get_prompt_modules(
                prepared.model_config.prompt_version
            )[1],
            {
                "current_volunteer": {"available_via_tool": True},
                "candidate_reference": {"available_via_tool": True},
                "current_job": {"available_via_tool": True},
                "resume_text": "available_via_readonly_tools",
            },
        )[0],
        model=KernelModelConfigV1(
            api_style=prepared.model_config.api_style,
            base_url=prepared.model_config.base_url,
            model_name=prepared.model_config.model_name,
            structured_output_mode=ai_config.get_structured_output_mode(
                api_style=prepared.model_config.api_style
            ),
            timeout_seconds=ai_config.get_ai_runtime_config().timeout_seconds,
            retry_count=ai_config.get_ai_runtime_config().retry_count,
            insecure_skip_verify=getattr(
                settings, "AGENT_KERNEL_MODEL_INSECURE_SKIP_VERIFY", False
            ),
        ),
    )
